### ISSUE TITLE
Experiment not running on Pavlovia

### ISSUE THREAD

**Issue by Khushbus21** [datalowe/psychopy-wm-spatial#1]
``hey, I tried running your experiment on Pavlovia online by syncing through PsychoPy, but it stuck on 'Initializing the Experiment' and the console gave this error -"Uncaught SyntaxError: Identifier 'Rect' has already been declared" in the line 55 of the compiled JS code

Can you please help me out

`/******************* 
 * Wm_Spatial Test *
 *******************/

import { PsychoJS } from './lib/core-2021.1.4.js';
import * as core from './lib/core-2021.1.4.js';
import { TrialHandler } from './lib/data-2021.1.4.js';
import { Scheduler } from './lib/util-2021.1.4.js';
import * as visual from './lib/visual-2021.1.4.js';
import * as sound from './lib/sound-2021.1.4.js';
import * as util from './lib/util-2021.1.4.js';
//some handy aliases as in the psychopy scripts;
const { abs, sin, cos, PI: pi, sqrt } = Math;
const { round } = util;

// init psychoJS:
const psychoJS = new PsychoJS({
  debug: true
});

// open window:
psychoJS.openWindow({
  fullscr: true,
  color: new util.Color([(- 1), (- 1), (- 1)]),
  units: 'deg',
  waitBlanking: true
});

// store info about the experiment session:
let expName = 'wm_spatial';  // from the Builder filename that created this script
let expInfo = {'participant': '', 'session': '001'};

// Start code blocks for 'Before Experiment'
AREA_WIDTH_DEG = 22;
AREA_HEIGHT_DEG = 13;
TARGET_SIZE_DEG = 2.5;
TARGET_NEUTRAL_COL = "#AAAAAA";
TARGET_LIGHT_COL = "#FFFF00";
TARGET_AWAIT_COL = "#FFFFFF";
TARGET_CLICKED_COL = "#BBFFBB";
NUM_TRIALS = 3;
NUM_TARGETS = 9;
START_DIFFICULTY = 2;
END_DIFFICULTY = 9;
DEMO_LIGHT_DUR = 60;
RESP_FLASH_DUR = 3;
TXT_SIZE_L = 2;
TXT_SIZE_M = 0.8;
TXT_SIZE_S = 0.6;
INSTRUCTIONS_TXT = "I det h\u00e4r testet kommer du att visas rutor. Rutorna kommer att lysa upp i en viss ordning.Du ska memorisera i vilken ordning som rutorna lyser upp. N\u00e4r rutorna lyst klart ska du klicka p\u00e5 rutorna i samma ordning.\n\nAnv\u00e4nd v\u00e4nster musknapp f\u00f6r att klicka p\u00e5 rutorna.";
INSTRUCTIONS_CONTINUE_TXT = "Tryck h\u00e4r f\u00f6r att se ett exempel";

import {Rect} from 'psychopy/visual/rect';

**_import {Rect} from 'psychopy/visual/rect';_**

// schedule the experiment:
psychoJS.schedule(psychoJS.gui.DlgFromDict({
  dictionary: expInfo,
  title: expName
}));

const flowScheduler = new Scheduler(psychoJS);
const dialogCancelScheduler = new Scheduler(psychoJS);
psychoJS.scheduleCondition(function() { return (psychoJS.gui.dialogComponent.button === 'OK'); }, flowScheduler, dialogCancelScheduler);

// flowScheduler gets run if the participants presses OK
flowScheduler.add(updateInfo); // add timeStamp
flowScheduler.add(experimentInit);
flowScheduler.add(instructions_part1RoutineBegin());
flowScheduler.add(instructions_part1RoutineEachFrame());
flowScheduler.add(instructions_part1RoutineEnd());
const practice_loopLoopScheduler = new Scheduler(psychoJS);
flowScheduler.add(practice_loopLoopBegin, practice_loopLoopScheduler);
flowScheduler.add(practice_loopLoopScheduler);
flowScheduler.add(practice_loopLoopEnd);
flowScheduler.add(instructions_part2RoutineBegin());
flowScheduler.add(instructions_part2RoutineEachFrame());
flowScheduler.add(instructions_part2RoutineEnd());
const trialsLoopScheduler = new Scheduler(psychoJS);
flowScheduler.add(trialsLoopBegin, trialsLoopScheduler);
flowScheduler.add(trialsLoopScheduler);
flowScheduler.add(trialsLoopEnd);
flowScheduler.add(end_routineRoutineBegin());
flowScheduler.add(end_routineRoutineEachFrame());
flowScheduler.add(end_routineRoutineEnd());
flowScheduler.add(quitPsychoJS, '', true);

// quit if user presses Cancel in dialog box:
dialogCancelScheduler.add(quitPsychoJS, '', false);

psychoJS.start({
  expName: expName,
  expInfo: expInfo,
  resources: [
  ]
});

psychoJS.experimentLogger.setLevel(core.Logger.ServerLevel.DEBUG);

var frameDur;
function updateInfo() {
  expInfo['date'] = util.MonotonicClock.getDateStr();  // add a simple timestamp
  expInfo['expName'] = expName;
  expInfo['psychopyVersion'] = '2021.1.4';
  expInfo['OS'] = window.navigator.platform;

  // store frame rate of monitor if we can measure it successfully
  expInfo['frameRate'] = psychoJS.window.getActualFrameRate();
  if (typeof expInfo['frameRate'] !== 'undefined')
    frameDur = 1.0 / Math.round(expInfo['frameRate']);
  else
    frameDur = 1.0 / 60.0; // couldn't get a reliable measure so guess

  // add info from the URL:
  util.addInfoFromUrl(expInfo);
  
  return Scheduler.Event.NEXT;
}

var instructions_part1Clock;
var text_instructions_1;
var mouse_instructions_1;
var text_continue;
var demo_practiceClock;
var text_practice;
var instructions_part2Clock;
var text_instructions_2;
var mouse_instructions_2;
var text_start;
var trialClock;
var mouse_trial;
var end_routineClock;
var text_end;
var globalClock;
var routineTimer;
function experimentInit() {
  // Initialize components for Routine "instructions_part1"
  instructions_part1Clock = new util.Clock();
  text_instructions_1 = new visual.TextStim({
    win: psychoJS.window,
    name: 'text_instructions_1',
    text: INSTRUCTIONS_TXT_1,
    font: 'Arial',
    units: 'deg', 
    pos: [0, 0], height: TXT_SIZE_M,  wrapWidth: 25, ori: 0,
    color: new util.Color('white'),  opacity: 1,
    depth: 0.0 
  });
  
  mouse_instructions_1 = new core.Mouse({
    win: psychoJS.window,
  });
  mouse_instructions_1.mouseClock = new util.Clock();
  text_continue = new visual.TextStim({
    win: psychoJS.window,
    name: 'text_continue',
    text: INSTRUCTIONS_CONTINUE_TXT,
    font: 'Arial',
    units: 'deg', 
    pos: INSTRUCTIONS_CONTINUE_TXT_PLACEMENT, height: TXT_SIZE_S,  wrapWidth: 25, ori: 0,
    color: new util.Color('white'),  opacity: 1,
    depth: -2.0 
  });
  
  // Initialize components for Routine "demo_practice"
  demo_practiceClock = new util.Clock();
  /* Syntax Error: Fix Python code */
  text_practice = new visual.TextStim({
    win: psychoJS.window,
    name: 'text_practice',
    text: PRACTICE_TXT,
    font: 'Arial',
    units: 'deg', 
    pos: PRACTICE_TXT_PLACEMENT, height: TXT_SIZE_M,  wrapWidth: 20, ori: 0,
    color: new util.Color('white'),  opacity: 1,
    depth: -1.0 
  });
  
  // Initialize components for Routine "instructions_part2"
  instructions_part2Clock = new util.Clock();
  text_instructions_2 = new visual.TextStim({
    win: psychoJS.window,
    name: 'text_instructions_2',
    text: INSTRUCTIONS_TXT_2,
    font: 'Arial',
    units: 'deg', 
    pos: [0, 0], height: TXT_SIZE_M,  wrapWidth: 25, ori: 0,
    color: new util.Color('white'),  opacity: 1,
    depth: 0.0 
  });
  
  mouse_instructions_2 = new core.Mouse({
    win: psychoJS.window,
  });
  mouse_instructions_2.mouseClock = new util.Clock();
  text_start = new visual.TextStim({
    win: psychoJS.window,
    name: 'text_start',
    text: INSTRUCTIONS_START_TXT,
    font: 'Arial',
    units: 'deg', 
    pos: INSTRUCTIONS_CONTINUE_TXT_PLACEMENT, height: TXT_SIZE_S,  wrapWidth: 25, ori: 0,
    color: new util.Color('white'),  opacity: 1,
    depth: -2.0 
  });
  
  // Initialize components for Routine "trial"
  trialClock = new util.Clock();
  /* Syntax Error: Fix Python code */
  mouse_trial = new core.Mouse({
    win: psychoJS.window,
  });
  mouse_trial.mouseClock = new util.Clock();
  // Initialize components for Routine "end_routine"
  end_routineClock = new util.Clock();
  text_end = new visual.TextStim({
    win: psychoJS.window,
    name: 'text_end',
    text: END_TXT,
    font: 'Arial',
    units: 'deg', 
    pos: [0, 0], height: TXT_SIZE_L,  wrapWidth: 25, ori: 0,
    color: new util.Color('white'),  opacity: 1,
    depth: 0.0 
  });
  
  // Create some handy timers
  globalClock = new util.Clock();  // to track the time since experiment started
  routineTimer = new util.CountdownTimer();  // to track time remaining of each (non-slip) routine
  
  return Scheduler.Event.NEXT;
}

var t;
var frameN;
var continueRoutine;
var gotValidClick;
var instructions_part1Components;
function instructions_part1RoutineBegin(snapshot) {
  return function () {
    //------Prepare to start Routine 'instructions_part1'-------
    t = 0;
    instructions_part1Clock.reset(); // clock
    frameN = -1;
    continueRoutine = true; // until we're told otherwise
    // update component parameters for each repeat
    // setup some python lists for storing info about the mouse_instructions_1
    mouse_instructions_1.clicked_name = [];
    gotValidClick = false; // until a click is received
    // keep track of which components have finished
    instructions_part1Components = [];
    instructions_part1Components.push(text_instructions_1);
    instructions_part1Components.push(mouse_instructions_1);
    instructions_part1Components.push(text_continue);
    
    for (const thisComponent of instructions_part1Components)
      if ('status' in thisComponent)
        thisComponent.status = PsychoJS.Status.NOT_STARTED;
    return Scheduler.Event.NEXT;
  }
}

var prevButtonState;
var _mouseButtons;
function instructions_part1RoutineEachFrame(snapshot) {
  return function () {
    //------Loop for each frame of Routine 'instructions_part1'-------
    // get current time
    t = instructions_part1Clock.getTime();
    frameN = frameN + 1;// number of completed frames (so 0 is the first frame)
    // update/draw components on each frame
    
    // *text_instructions_1* updates
    if (t >= 0.0 && text_instructions_1.status === PsychoJS.Status.NOT_STARTED) {
      // keep track of start time/frame for later
      text_instructions_1.tStart = t;  // (not accounting for frame time here)
      text_instructions_1.frameNStart = frameN;  // exact frame index
      
      text_instructions_1.setAutoDraw(true);
    }

    // *mouse_instructions_1* updates
    if (t >= 0.5 && mouse_instructions_1.status === PsychoJS.Status.NOT_STARTED) {
      // keep track of start time/frame for later
      mouse_instructions_1.tStart = t;  // (not accounting for frame time here)
      mouse_instructions_1.frameNStart = frameN;  // exact frame index
      
      mouse_instructions_1.status = PsychoJS.Status.STARTED;
      mouse_instructions_1.mouseClock.reset();
      prevButtonState = mouse_instructions_1.getPressed();  // if button is down already this ISN'T a new click
      }
    if (mouse_instructions_1.status === PsychoJS.Status.STARTED) {  // only update if started and not finished!
      _mouseButtons = mouse_instructions_1.getPressed();
      if (!_mouseButtons.every( (e,i,) => (e == prevButtonState[i]) )) { // button state changed?
        prevButtonState = _mouseButtons;
        if (_mouseButtons.reduce( (e, acc) => (e+acc) ) > 0) { // state changed to a new click
          // check if the mouse was inside our 'clickable' objects
          gotValidClick = false;
          for (const obj of [text_continue]) {
            if (obj.contains(mouse_instructions_1)) {
              gotValidClick = true;
              mouse_instructions_1.clicked_name.push(obj.name)
            }
          }
          if (gotValidClick === true) { // abort routine on response
            continueRoutine = false;
          }
        }
      }
    }
    
    // *text_continue* updates
    if (t >= 0.0 && text_continue.status === PsychoJS.Status.NOT_STARTED) {
      // keep track of start time/frame for later
      text_continue.tStart = t;  // (not accounting for frame time here)
      text_continue.frameNStart = frameN;  // exact frame index
      
      text_continue.setAutoDraw(true);
    }

    // check for quit (typically the Esc key)
    if (psychoJS.experiment.experimentEnded || psychoJS.eventManager.getKeys({keyList:['escape']}).length > 0) {
      return quitPsychoJS('The [Escape] key was pressed. Goodbye!', false);
    }
    
    // check if the Routine should terminate
    if (!continueRoutine) {  // a component has requested a forced-end of Routine
      return Scheduler.Event.NEXT;
    }
    
    continueRoutine = false;  // reverts to True if at least one component still running
    for (const thisComponent of instructions_part1Components)
      if ('status' in thisComponent && thisComponent.status !== PsychoJS.Status.FINISHED) {
        continueRoutine = true;
        break;
      }
    
    // refresh the screen if continuing
    if (continueRoutine) {
      return Scheduler.Event.FLIP_REPEAT;
    } else {
      return Scheduler.Event.NEXT;
    }
  };
}

function instructions_part1RoutineEnd(snapshot) {
  return function () {
    //------Ending Routine 'instructions_part1'-------
    for (const thisComponent of instructions_part1Components) {
      if (typeof thisComponent.setAutoDraw === 'function') {
        thisComponent.setAutoDraw(false);
      }
    }
    // store data for thisExp (ExperimentHandler)
    // the Routine "instructions_part1" was not non-slip safe, so reset the non-slip timer
    routineTimer.reset();
    
    return Scheduler.Event.NEXT;
  };
}

var practice_loop;
var currentLoop;
function practice_loopLoopBegin(practice_loopLoopScheduler) {
  // set up handler to look after randomisation of conditions etc
  practice_loop = new TrialHandler({
    psychoJS: psychoJS,
    nReps: 9999, method: TrialHandler.Method.RANDOM,
    extraInfo: expInfo, originPath: undefined,
    trialList: undefined,
    seed: undefined, name: 'practice_loop'
  });
  psychoJS.experiment.addLoop(practice_loop); // add the loop to the experiment
  currentLoop = practice_loop;  // we're now the current loop

  // Schedule all the trials in the trialList:
  for (const thisPractice_loop of practice_loop) {
    const snapshot = practice_loop.getSnapshot();
    practice_loopLoopScheduler.add(importConditions(snapshot));
    practice_loopLoopScheduler.add(demo_practiceRoutineBegin(snapshot));
    practice_loopLoopScheduler.add(demo_practiceRoutineEachFrame(snapshot));
    practice_loopLoopScheduler.add(demo_practiceRoutineEnd(snapshot));
    practice_loopLoopScheduler.add(endLoopIteration(practice_loopLoopScheduler, snapshot));
  }

  return Scheduler.Event.NEXT;
}

function practice_loopLoopEnd() {
  psychoJS.experiment.removeLoop(practice_loop);

  return Scheduler.Event.NEXT;
}

var trials;
function trialsLoopBegin(trialsLoopScheduler) {
  // set up handler to look after randomisation of conditions etc
  trials = new TrialHandler({
    psychoJS: psychoJS,
    nReps: NUM_TRIALS * (END_DIFFICULTY - START_DIFFICULTY + 1), method: TrialHandler.Method.RANDOM,
    extraInfo: expInfo, originPath: undefined,
    trialList: undefined,
    seed: undefined, name: 'trials'
  });
  psychoJS.experiment.addLoop(trials); // add the loop to the experiment
  currentLoop = trials;  // we're now the current loop

  // Schedule all the trials in the trialList:
  for (const thisTrial of trials) {
    const snapshot = trials.getSnapshot();
    trialsLoopScheduler.add(importConditions(snapshot));
    trialsLoopScheduler.add(trialRoutineBegin(snapshot));
    trialsLoopScheduler.add(trialRoutineEachFrame(snapshot));
    trialsLoopScheduler.add(trialRoutineEnd(snapshot));
    trialsLoopScheduler.add(endLoopIteration(trialsLoopScheduler, snapshot));
  }

  return Scheduler.Event.NEXT;
}

function trialsLoopEnd() {
  psychoJS.experiment.removeLoop(trials);

  return Scheduler.Event.NEXT;
}

var light_order;
var light_targets;
var seq_len;
var light_counter;
var light_target;
var response_phase;
var click_counter;
var response_times;
var click_order;
var trial_start_time;
var demo_practiceComponents;
function demo_practiceRoutineBegin(snapshot) {
  return function () {
    //------Prepare to start Routine 'demo_practice'-------
    t = 0;
    demo_practiceClock.reset(); // clock
    frameN = -1;
    continueRoutine = true; // until we're told otherwise
    // update component parameters for each repeat
    light_order = trial_orders[trial_counter];
    light_targets = function () {
        var _pj_a = [], _pj_b = light_order;
        for (var _pj_c = 0, _pj_d = _pj_b.length; (_pj_c < _pj_d); _pj_c += 1) {
            var index = _pj_b[_pj_c];
            _pj_a.push(targets[index]);
        }
        return _pj_a;
    }
    .call(this);
    seq_len = light_targets.length;
    light_counter = 0;
    light_target = light_targets[light_counter];
    light_target.switch_on(60);
    response_phase = false;
    click_counter = 0;
    response_times = [];
    click_order = [];
    trial_start_time = globalClock.getTime();
    
    // keep track of which components have finished
    demo_practiceComponents = [];
    demo_practiceComponents.push(text_practice);
    
    for (const thisComponent of demo_practiceComponents)
      if ('status' in thisComponent)
        thisComponent.status = PsychoJS.Status.NOT_STARTED;
    return Scheduler.Event.NEXT;
  }
}

function demo_practiceRoutineEachFrame(snapshot) {
  return function () {
    //------Loop for each frame of Routine 'demo_practice'-------
    // get current time
    t = demo_practiceClock.getTime();
    frameN = frameN + 1;// number of completed frames (so 0 is the first frame)
    // update/draw components on each frame
    /* Syntax Error: Fix Python code */
    
    // *text_practice* updates
    if (t >= 9999 && text_practice.status === PsychoJS.Status.NOT_STARTED) {
      // keep track of start time/frame for later
      text_practice.tStart = t;  // (not accounting for frame time here)
      text_practice.frameNStart = frameN;  // exact frame index
      
      text_practice.setAutoDraw(true);
    }

    // check for quit (typically the Esc key)
    if (psychoJS.experiment.experimentEnded || psychoJS.eventManager.getKeys({keyList:['escape']}).length > 0) {
      return quitPsychoJS('The [Escape] key was pressed. Goodbye!', false);
    }
    
    // check if the Routine should terminate
    if (!continueRoutine) {  // a component has requested a forced-end of Routine
      return Scheduler.Event.NEXT;
    }
    
    continueRoutine = false;  // reverts to True if at least one component still running
    for (const thisComponent of demo_practiceComponents)
      if ('status' in thisComponent && thisComponent.status !== PsychoJS.Status.FINISHED) {
        continueRoutine = true;
        break;
      }
    
    // refresh the screen if continuing
    if (continueRoutine) {
      return Scheduler.Event.FLIP_REPEAT;
    } else {
      return Scheduler.Event.NEXT;
    }
  };
}

function demo_practiceRoutineEnd(snapshot) {
  return function () {
    //------Ending Routine 'demo_practice'-------
    for (const thisComponent of demo_practiceComponents) {
      if (typeof thisComponent.setAutoDraw === 'function') {
        thisComponent.setAutoDraw(false);
      }
    }
    trial_counter += 1;
    
    // the Routine "demo_practice" was not non-slip safe, so reset the non-slip timer
    routineTimer.reset();
    
    return Scheduler.Event.NEXT;
  };
}

var instructions_part2Components;
function instructions_part2RoutineBegin(snapshot) {
  return function () {
    //------Prepare to start Routine 'instructions_part2'-------
    t = 0;
    instructions_part2Clock.reset(); // clock
    frameN = -1;
    continueRoutine = true; // until we're told otherwise
    // update component parameters for each repeat
    // setup some python lists for storing info about the mouse_instructions_2
    mouse_instructions_2.clicked_name = [];
    gotValidClick = false; // until a click is received
    // keep track of which components have finished
    instructions_part2Components = [];
    instructions_part2Components.push(text_instructions_2);
    instructions_part2Components.push(mouse_instructions_2);
    instructions_part2Components.push(text_start);
    
    for (const thisComponent of instructions_part2Components)
      if ('status' in thisComponent)
        thisComponent.status = PsychoJS.Status.NOT_STARTED;
    return Scheduler.Event.NEXT;
  }
}

function instructions_part2RoutineEachFrame(snapshot) {
  return function () {
    //------Loop for each frame of Routine 'instructions_part2'-------
    // get current time
    t = instructions_part2Clock.getTime();
    frameN = frameN + 1;// number of completed frames (so 0 is the first frame)
    // update/draw components on each frame
    
    // *text_instructions_2* updates
    if (t >= 0.0 && text_instructions_2.status === PsychoJS.Status.NOT_STARTED) {
      // keep track of start time/frame for later
      text_instructions_2.tStart = t;  // (not accounting for frame time here)
      text_instructions_2.frameNStart = frameN;  // exact frame index
      
      text_instructions_2.setAutoDraw(true);
    }

    // *mouse_instructions_2* updates
    if (t >= 0.5 && mouse_instructions_2.status === PsychoJS.Status.NOT_STARTED) {
      // keep track of start time/frame for later
      mouse_instructions_2.tStart = t;  // (not accounting for frame time here)
      mouse_instructions_2.frameNStart = frameN;  // exact frame index
      
      mouse_instructions_2.status = PsychoJS.Status.STARTED;
      mouse_instructions_2.mouseClock.reset();
      prevButtonState = mouse_instructions_2.getPressed();  // if button is down already this ISN'T a new click
      }
    if (mouse_instructions_2.status === PsychoJS.Status.STARTED) {  // only update if started and not finished!
      _mouseButtons = mouse_instructions_2.getPressed();
      if (!_mouseButtons.every( (e,i,) => (e == prevButtonState[i]) )) { // button state changed?
        prevButtonState = _mouseButtons;
        if (_mouseButtons.reduce( (e, acc) => (e+acc) ) > 0) { // state changed to a new click
          // check if the mouse was inside our 'clickable' objects
          gotValidClick = false;
          for (const obj of [text_start]) {
            if (obj.contains(mouse_instructions_2)) {
              gotValidClick = true;
              mouse_instructions_2.clicked_name.push(obj.name)
            }
          }
          if (gotValidClick === true) { // abort routine on response
            continueRoutine = false;
          }
        }
      }
    }
    
    // *text_start* updates
    if (t >= 0.0 && text_start.status === PsychoJS.Status.NOT_STARTED) {
      // keep track of start time/frame for later
      text_start.tStart = t;  // (not accounting for frame time here)
      text_start.frameNStart = frameN;  // exact frame index
      
      text_start.setAutoDraw(true);
    }

    // check for quit (typically the Esc key)
    if (psychoJS.experiment.experimentEnded || psychoJS.eventManager.getKeys({keyList:['escape']}).length > 0) {
      return quitPsychoJS('The [Escape] key was pressed. Goodbye!', false);
    }
    
    // check if the Routine should terminate
    if (!continueRoutine) {  // a component has requested a forced-end of Routine
      return Scheduler.Event.NEXT;
    }
    
    continueRoutine = false;  // reverts to True if at least one component still running
    for (const thisComponent of instructions_part2Components)
      if ('status' in thisComponent && thisComponent.status !== PsychoJS.Status.FINISHED) {
        continueRoutine = true;
        break;
      }
    
    // refresh the screen if continuing
    if (continueRoutine) {
      return Scheduler.Event.FLIP_REPEAT;
    } else {
      return Scheduler.Event.NEXT;
    }
  };
}

function instructions_part2RoutineEnd(snapshot) {
  return function () {
    //------Ending Routine 'instructions_part2'-------
    for (const thisComponent of instructions_part2Components) {
      if (typeof thisComponent.setAutoDraw === 'function') {
        thisComponent.setAutoDraw(false);
      }
    }
    // store data for thisExp (ExperimentHandler)
    // the Routine "instructions_part2" was not non-slip safe, so reset the non-slip timer
    routineTimer.reset();
    
    return Scheduler.Event.NEXT;
  };
}

var trialComponents;
function trialRoutineBegin(snapshot) {
  return function () {
    //------Prepare to start Routine 'trial'-------
    t = 0;
    trialClock.reset(); // clock
    frameN = -1;
    continueRoutine = true; // until we're told otherwise
    // update component parameters for each repeat
    light_order = trial_orders[trial_counter];
    light_targets = function () {
        var _pj_a = [], _pj_b = light_order;
        for (var _pj_c = 0, _pj_d = _pj_b.length; (_pj_c < _pj_d); _pj_c += 1) {
            var index = _pj_b[_pj_c];
            _pj_a.push(targets[index]);
        }
        return _pj_a;
    }
    .call(this);
    seq_len = light_targets.length;
    light_counter = 0;
    light_target = light_targets[light_counter];
    light_target.switch_on(60);
    response_phase = false;
    click_counter = 0;
    response_times = [];
    click_order = [];
    trial_start_time = globalClock.getTime();
    
    // setup some python lists for storing info about the mouse_trial
    gotValidClick = false; // until a click is received
    // keep track of which components have finished
    trialComponents = [];
    trialComponents.push(mouse_trial);
    
    for (const thisComponent of trialComponents)
      if ('status' in thisComponent)
        thisComponent.status = PsychoJS.Status.NOT_STARTED;
    return Scheduler.Event.NEXT;
  }
}

function trialRoutineEachFrame(snapshot) {
  return function () {
    //------Loop for each frame of Routine 'trial'-------
    // get current time
    t = trialClock.getTime();
    frameN = frameN + 1;// number of completed frames (so 0 is the first frame)
    // update/draw components on each frame
    /* Syntax Error: Fix Python code */
    // check for quit (typically the Esc key)
    if (psychoJS.experiment.experimentEnded || psychoJS.eventManager.getKeys({keyList:['escape']}).length > 0) {
      return quitPsychoJS('The [Escape] key was pressed. Goodbye!', false);
    }
    
    // check if the Routine should terminate
    if (!continueRoutine) {  // a component has requested a forced-end of Routine
      return Scheduler.Event.NEXT;
    }
    
    continueRoutine = false;  // reverts to True if at least one component still running
    for (const thisComponent of trialComponents)
      if ('status' in thisComponent && thisComponent.status !== PsychoJS.Status.FINISHED) {
        continueRoutine = true;
        break;
      }
    
    // refresh the screen if continuing
    if (continueRoutine) {
      return Scheduler.Event.FLIP_REPEAT;
    } else {
      return Scheduler.Event.NEXT;
    }
  };
}

function trialRoutineEnd(snapshot) {
  return function () {
    //------Ending Routine 'trial'-------
    for (const thisComponent of trialComponents) {
      if (typeof thisComponent.setAutoDraw === 'function') {
        thisComponent.setAutoDraw(false);
      }
    }
    trial_counter += 1;
    
    // store data for thisExp (ExperimentHandler)
    // the Routine "trial" was not non-slip safe, so reset the non-slip timer
    routineTimer.reset();
    
    return Scheduler.Event.NEXT;
  };
}

var end_routineComponents;
function end_routineRoutineBegin(snapshot) {
  return function () {
    //------Prepare to start Routine 'end_routine'-------
    t = 0;
    end_routineClock.reset(); // clock
    frameN = -1;
    continueRoutine = true; // until we're told otherwise
    routineTimer.add(5.000000);
    // update component parameters for each repeat
    // keep track of which components have finished
    end_routineComponents = [];
    end_routineComponents.push(text_end);
    
    for (const thisComponent of end_routineComponents)
      if ('status' in thisComponent)
        thisComponent.status = PsychoJS.Status.NOT_STARTED;
    return Scheduler.Event.NEXT;
  }
}

var frameRemains;
function end_routineRoutineEachFrame(snapshot) {
  return function () {
    //------Loop for each frame of Routine 'end_routine'-------
    // get current time
    t = end_routineClock.getTime();
    frameN = frameN + 1;// number of completed frames (so 0 is the first frame)
    // update/draw components on each frame
    
    // *text_end* updates
    if (t >= 0.0 && text_end.status === PsychoJS.Status.NOT_STARTED) {
      // keep track of start time/frame for later
      text_end.tStart = t;  // (not accounting for frame time here)
      text_end.frameNStart = frameN;  // exact frame index
      
      text_end.setAutoDraw(true);
    }

    frameRemains = 0.0 + 5 - psychoJS.window.monitorFramePeriod * 0.75;  // most of one frame period left
    if (text_end.status === PsychoJS.Status.STARTED && t >= frameRemains) {
      text_end.setAutoDraw(false);
    }
    // check for quit (typically the Esc key)
    if (psychoJS.experiment.experimentEnded || psychoJS.eventManager.getKeys({keyList:['escape']}).length > 0) {
      return quitPsychoJS('The [Escape] key was pressed. Goodbye!', false);
    }
    
    // check if the Routine should terminate
    if (!continueRoutine) {  // a component has requested a forced-end of Routine
      return Scheduler.Event.NEXT;
    }
    
    continueRoutine = false;  // reverts to True if at least one component still running
    for (const thisComponent of end_routineComponents)
      if ('status' in thisComponent && thisComponent.status !== PsychoJS.Status.FINISHED) {
        continueRoutine = true;
        break;
      }
    
    // refresh the screen if continuing
    if (continueRoutine && routineTimer.getTime() > 0) {
      return Scheduler.Event.FLIP_REPEAT;
    } else {
      return Scheduler.Event.NEXT;
    }
  };
}

function end_routineRoutineEnd(snapshot) {
  return function () {
    //------Ending Routine 'end_routine'-------
    for (const thisComponent of end_routineComponents) {
      if (typeof thisComponent.setAutoDraw === 'function') {
        thisComponent.setAutoDraw(false);
      }
    }
    return Scheduler.Event.NEXT;
  };
}

function endLoopIteration(scheduler, snapshot) {
  // ------Prepare for next entry------
  return function () {
    if (typeof snapshot !== 'undefined') {
      // ------Check if user ended loop early------
      if (snapshot.finished) {
        // Check for and save orphaned data
        if (psychoJS.experiment.isEntryEmpty()) {
          psychoJS.experiment.nextEntry(snapshot);
        }
        scheduler.stop();
      } else {
        const thisTrial = snapshot.getCurrentTrial();
        if (typeof thisTrial === 'undefined' || !('isTrials' in thisTrial) || thisTrial.isTrials) {
          psychoJS.experiment.nextEntry(snapshot);
        }
      }
    return Scheduler.Event.NEXT;
    }
  };
}

function importConditions(currentLoop) {
  return function () {
    psychoJS.importAttributes(currentLoop.getCurrentTrial());
    return Scheduler.Event.NEXT;
    };
}

function quitPsychoJS(message, isCompleted) {
  // Check for and save orphaned data
  if (psychoJS.experiment.isEntryEmpty()) {
    psychoJS.experiment.nextEntry();
  }
  
  
  
  
  
  
  psychoJS.window.close();
  psychoJS.quit({message: message, isCompleted: isCompleted});
  
  return Scheduler.Event.QUIT;
}
`

**Comment by datalowe** [comment]
Hi, thanks for raising this issue :)

I actually haven't used Pavlovia/PsychoPy's JavaScript library myself. This should be mentioned in the README. Since the Builder project includes a lot of custom code snippets, it's likely that the automatic transpilation ('translation') from Python to JavaScript will run into multiple issues.

This particular issue relates to how an explicit import needs to be included in the Python code to circumvent a bug in PsychoPy's own Python libraries:
https://github.com/psychopy/psychopy/issues/1159

For this reason, one of this project's Builder code snippets includes:
```py
from psychopy.visual.rect import Rect
```
Based on the log output you've provided, it seems like, for whatever reason, the automatic transpiler ('code translator') duplicates this import statement. Perhaps it's a matter of the transpiler trying to be 'clever' and automatically adding its own imports of classes that are directly referenced in code, not taking into account that the developer might have added an explicit import already. This seems, at face value at least, like a bug in the transpiler. 

It's very likely that one would have to go through and 'help out' the transpiler by editing the generated JavaScript code directly, to help it solve this and presumably many other issues. It might be a long time before I have time for this, unfortunately. 

If you want to give 'helping the transpiler' a try then please fork this project. And if you're able to make it work, you're very welcome to make a pull request. Do note that this might be time-consuming unless you have a lot of experience with PsychoPy and Javascript (or if the transpilation works better than I suspect). If you don't, I'd only recommend doing this if you're interested in it as a learning exercise.

**Comment by datalowe** [comment]
I made an honest attempt at making the experiment 'transpilation-friendly', even trying to manually translate some of the code snippets by hand. Unfortunately, even trying to translate by hand seems like it might not be enough, since 'psychojs' makes some assumptions about namespacing that the Python implementation of PsychoPy doesn't. I've added a note to the README warning users who might attempt a JS translation.
